soroban-sdk/src/contract_event.rs

```diff
@@ -24,43 +32,91 @@ use crate::{env::internal, Env, IntoVal, RawVal, Vec};
 /// # #[cfg(feature = "testutils")]
 /// # fn main() {
 /// #     let env = Env::default();
-/// #     let contract_id = FixedBinary::from_array(&env, [0; 32]);
+/// #     let contract_id = BytesN::from_array(&env, [0; 32]);
 /// #     env.register_contract(&contract_id, Contract);
 /// #     f::invoke(&env, &contract_id);
 /// # }
 /// # #[cfg(not(feature = "testutils"))]
 /// # fn main() { }
 /// ```
 #[derive(Clone)]
-pub struct ContractEvent(Env);
+pub struct Events(Env);
 
-impl Debug for ContractEvent {
+impl Debug for Events {
     fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
-        write!(f, "ContractEvent")
+        write!(f, "Events")
+    }
+}
+
+pub trait IntoTopics {
+    fn into_topics(self, env: &Env) -> Object;
+}
+
+// 0 topics
+impl IntoTopics for () {
+    fn into_topics(self, env: &Env) -> Object {
+        Vec::<RawVal>::new(env).to_object()
     }
 }
 
-impl ContractEvent {
+macro_rules! impl_for_tuple {
+    ( $($typ:ident $idx:tt)* ) => {
+        impl<$($typ),*> IntoTopics for ($($typ,)*)
+        where
+            $($typ: IntoVal<Env, RawVal> + Clone),*
+        {
+            fn into_topics(self, env: &Env) -> Object {
+                $({
+                    let ev = self.$idx.clone().into_env_val(env);
+                    match TryInto::<EnvObj>::try_into(ev) {
+                        Ok(obj) => {
+                            if obj.as_object().is_obj_type(ScObjectType::Vec) {
+                                panic!("topic cannot be a vec")
+                            } else if obj.as_object().is_obj_type(ScObjectType::Map) {
+                                panic!("topic cannot be a map")
+                            } else if obj.as_object().is_obj_type(ScObjectType::Bytes) {
+                                let bytes = unsafe { Bytes::unchecked_new(obj.clone()) };
+                                if bytes.len() > TOPIC_BYTES_LENGTH_LIMIT {
+                                    panic!("topic exceeds bytes length limit")
+                                }
+                            }
+                        }
+                        Err(_) => (),
+                    }
+                })*
+                self.into_val(env)
+            }
+        }
+    };
+}
+
+// 1-4 topics
+impl_for_tuple! { T0 0 }
+impl_for_tuple! { T0 0 T1 1 }
+impl_for_tuple! { T0 0 T1 1 T2 2 }
+impl_for_tuple! { T0 0 T1 1 T2 2 T3 3 }
+
+impl Events {
     #[inline(always)]
     pub(crate) fn env(&self) -> &Env {
         &self.0
     }
 
     #[inline(always)]
-    pub(crate) fn new(env: &Env) -> ContractEvent {
-        ContractEvent(env.clone())
+    pub(crate) fn new(env: &Env) -> Events {
+        Events(env.clone())
     }
 
     /// Publishes `topics` and `data` into a contract event.
     /// `topics` is expected to have length <= 4 and cannot contain Vecs, Maps,
     /// or Binaries > 32 bytes
```

Comment:
Keep the first line of the docs short and brief because it is rendered in places in the docs separate to the rest.

```suggestion
     /// Publish an event.
     ///
     /// Event data is specified in `data`. Data may be any value or
     /// type, including types defined by contracts using [contracttype].
```

Reply:
For the above doc comment you'll need to also include this in the code:
```rust
 #[cfg(doc)]
 use crate::{contracttype, Bytes, BytesN, Map};
```

Reply:
To clarify, `contracttype` *can* be a topic? It is listed in the "must not contain" category in your comment.

Reply:
No, can't be, because contract types are Vecs and Maps.

Reply:
Thanks, updated the comment.